pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go

```diff
@@ -652,39 +658,57 @@ func (lrw *logicalReplicationWriterProcessor) flushBuffer(
 		return b.checkpoint, nil
 	}
 
+	batchSize := int(flushBatchSize.Get(&lrw.EvalCtx.Settings.SV))
+
 	// Ensure the batcher is always reset, even on early error returns.
 	preFlushTime := timeutil.Now()
 	lrw.debug.RecordFlushStart(preFlushTime, int64(len(b.buffer.curKVBatch)))
 
+	workers := min(maxWriterWorkers, (len(b.buffer.curKVBatch)/batchSize)+1)
+	perWorker := (len(b.buffer.curKVBatch) / workers) + 1
 	// TODO: The batching here in production would need to be much
 	// smarter. Namely, we don't want to include updates to the
 	// same key in the same batch. Also, it's possible batching
 	// will make things much worse in practice.
-	flushByteSize := 0
-	batchStart := 0
-	batchSize := int(flushBatchSize.Get(&lrw.EvalCtx.Settings.SV))
-	batchEnd := min(batchStart+batchSize, len(b.buffer.curKVBatch))
-	for batchStart < len(b.buffer.curKVBatch) && batchEnd != 0 {
-		preBatchTime := timeutil.Now()
-		batchStats, err := lrw.bh.HandleBatch(ctx, b.buffer.curKVBatch[batchStart:batchEnd])
-		if err != nil {
-			// TODO(ssd): Handle errors. We should perhaps split the batch and retry a portion of the batch.
-			// If that fails, send the failed application to the dead-letter-queue.
-			return nil, err
+	var flushByteSize atomic.Int64
+
+	g := ctxgroup.WithContext(ctx)
+	for i := 0; i < workers; i++ {
+		bh := lrw.bh[i]
+		batchStart := i * perWorker
+		chunkEnd := batchStart + perWorker
+		if i == workers-1 {
+			chunkEnd = len(b.buffer.curKVBatch)
 		}
 
-		flushByteSize += batchStats.byteSize
-		batchStart = batchEnd
+		g.GoCtx(func(ctx context.Context) error {
+			for batchStart < chunkEnd {
+				batchEnd := min(batchStart+batchSize, chunkEnd)
+				preBatchTime := timeutil.Now()
+				batchStats, err := bh.HandleBatch(ctx, b.buffer.curKVBatch[batchStart:batchEnd])
+				if err != nil {
+					// TODO(ssd): Handle errors. We should perhaps split the batch and retry a portion of the batch.
+					// If that fails, send the failed application to the dead-letter-queue.
+					return err
+				}
+				batchStart = batchEnd
+				batchTime := timeutil.Since(preBatchTime)
+
+				lrw.debug.RecordBatchApplied(batchTime, int64(batchEnd-batchStart))
+				lrw.metrics.BatchBytesHist.RecordValue(int64(batchStats.byteSize))
+				lrw.metrics.BatchHistNanos.RecordValue(batchTime.Nanoseconds())
+				flushByteSize.Add(int64(batchStats.byteSize))
```

Comment:
It probably doesn't matter but we could have a local counter and then only update flushByteSize atomically at the end of the loop.

Reply:
i actually prefer incrementing stats per batch. incrementing per flush can lead to odd metrics graphs, seen here: https://cockroachlabs.slack.com/archives/C077F53B4KY/p1718314619407909?thread_ts=1718314538.225149&cid=C077F53B4KY

Reply:
mind actually incrementing  per batch as well?

```
lrw.metrics.IngestedLogicalBytes.Inc(byteCount)
lrw.metrics.CommitLatency.RecordValue(timeutil.Since(b.buffer.minTimestamp.GoTime()).Nanoseconds())
lrw.metrics.IngestedEvents.Inc(int64(len(b.buffer.curKVBatch)))
```

Reply:
> incrementing stats per batch.
I thought this too Bach when the flush took minutes;  with this change it os fast enough you rarely observe mid-flush

Reply:
I think changes to what metrics we change where should be a separate change.